keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1428 @@
+  provisioning a key and certificate chain on behalf of a pod.
+Taken together, this machinery makes it feasible to securely and automatically
+deliver X.509 certificates to every pod in a cluster, without imposing an
+unreasonable burden on application developers or cluster administrators.  
+
+As a first application of this machinery, this KEP defines additional machinery
+that allows pods to automatically receive mTLS client certificates suitable for
+authenticating to kube-apiserver, with the same attributes provided by bound
+service account tokens:
+* A PodIdentity X.509 extension to allow an X.509 certificate to carry all
+  details of a pod's identity, including its UIDs and the node it is running on.
+* A new TLS user-mapping capability in kube-apiserver to understand client
+  certificates that embed PodIdentity extensions.
+* A new PodCertificateRequest signer
+  (`kubernetes.io/kube-apiserver-client-pod`), shipped in
+  kube-controller-manager, to issue compatible client certificate to pods.
+* An enhancement to client-go's InClusterConfig that automatically loads a
+  credential bundle from a well-known filesystem path and uses it as a client
+  certificate.
+
+While this first application motivates some of the design decisions in this KEP,
+it is explicitly intended that PodCertificateRequests and PodCertificate
+projected volume sources will be useful to third-party projects that want to
+easily and securely issue X.509 certificates to workloads running in a
+Kubernetes cluster.
+
+There is a draft implementation of the design outlined in this KEP:
+https://github.com/kubernetes/kubernetes/pull/127251
+
+## Motivation
+
+### Goals
+
+**Ensure that certificate issuance can be accomplished without relying on bearer
+tokens at any step**:  One of the most significant benefits of mTLS
+authentication is that the credential is split into two pieces (private key and
+certificate), only one of which is ever transmitted to the counterparty (and
+thus subject to eavesdropping or theft by the counterparty).  In order to
+preserve this guarantee, it's important that bearer tokens are not involved *at
+any step*, otherwise the security of the scheme devolves to just being a bearer
+token.  In this KEP, as long as both kubelet and the signer implementation
+authenticate to kube-apiserver using their own separately-bootstrapped client
+certificates, then no bearer tokens are involved in certificate issuance.
+
+**Make it easy to ship secure third-party signer implementations**:  Securely
+implementing and operating a third-party signer implementation today requires
+intimate familiarity with Kubernetes' security model, otherwise your signer may
+negate Kubernetes' node restriction security boundary.  This KEP moves
+responsibility for enforcing node restriction on PodCertificateRequests into
+kube-apiserver, so that signer implementations do not need to consider it.
+
+### Non-Goals
```

Comment:
i think it's worth mentioning explicitly as a non-goal that this proposal does not cover signing serving certificates for pods. that seems obvious and covered in the pod-to-pod mTLS point, but implicitly.

Reply:
It's a non-goal for the built-in signer to issue serving certs (although we will probably mark the certs to allow server cert usage; it doesn't hurt anything).

However, it's an explicit goal for third-party signers to be able to use the PodCertificate projected volume machinery to be able to issue serving certificates to pods.

Reply:
Clarified in the text.

Reply:
> I would not expect us to mark these certs as anything but client certs.

Agreed.  As discussed on slack, we will only set the client cert usages.  This doesn't preclude users from using them as server certs; since they will need to write custom TLS validation code, they can just ignore the lack of the "server" usage bits.